LemMinX crashes if client configuration contains "xml.validation.schema"

https://github.com/eclipse/lemminx/commit/6e5a9061fec3c90bc49d64ad227f23cd0a23f28e changed the configuration model from

```
"xml.validation.schema": true,
```

to

```
"xml.validation.schema.enabled": "always",
```

If a client still sends the old setting, LemMinX throws a not too easy to resolve exception and that's it.

<details>

```xml
Message: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "workspace/didChangeConfiguration",
  "params": {
    "settings": {
      "xml": {
...
        "validation": {
          "noGrammar": "hint",
          "resolveExternalEntities": false,
          "disallowDocTypeDecl": false,
          "schema": true,
          "enabled": true
        },
...

java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was BOOLEAN at path $.validation.schema
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.Gson.fromJson(Gson.java:888)
	at com.google.gson.Gson.fromJson(Gson.java:953)
	at com.google.gson.Gson.fromJson(Gson.java:926)
	at org.eclipse.lemminx.utils.JSONUtility.toModel(JSONUtility.java:38)
	at org.eclipse.lemminx.extensions.contentmodel.settings.ContentModelSettings.getContentModelXMLSettings(ContentModelSettings.java:88)
	at org.eclipse.lemminx.XMLLanguageServer.updateSettings(XMLLanguageServer.java:207)
	at org.eclipse.lemminx.XMLLanguageServer.updateSettings(XMLLanguageServer.java:153)
	at org.eclipse.lemminx.XMLWorkspaceService.didChangeConfiguration(XMLWorkspaceService.java:81)
	... 16 more
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was BOOLEAN at path $.validation.schema
	at com.google.gson.internal.bind.JsonTreeReader.expect(JsonTreeReader.java:162)
	at com.google.gson.internal.bind.JsonTreeReader.beginObject(JsonTreeReader.java:87)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:213)
	... 28 more
```
</details>


This behavior might cause a not as smooth as expected transition to the new version.

The exception should be catched and transformed into a valuable error response with some hints about the changed settings model.

--- a/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/settings/XMLValidationSettings.java
+++ b/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/settings/XMLValidationSettings.java
@@ -45,11 +45,13 @@ public class XMLValidationSettings {
 		setEnabled(true);
 		setDisallowDocTypeDecl(false);
 		setResolveExternalEntities(false);
+		setNamespaces(new XMLNamespacesSettings());
+		setSchema(new XMLSchemaSettings());
 	}
 
 	/**
 	 * Returns true if the validation is enabled and false otherwise.
-	 * 
+	 *
 	 * @return true if the validation is enabled and false otherwise.
 	 */
 	public boolean isEnabled() {
@@ -58,7 +60,7 @@ public class XMLValidationSettings {
 
 	/**
 	 * Set true if the validation is enabled and false otherwise.
-	 * 
+	 *
 	 * @param enabled true if the validation is enabled and false otherwise.
 	 */
 	public void setEnabled(boolean enabled) {
@@ -67,7 +69,7 @@ public class XMLValidationSettings {
 
 	/**
 	 * Returns the XML Namespaces validation settings.
-	 * 
+	 *
 	 * @return the XML Namespaces validation settings.
 	 */
 	public XMLNamespacesSettings getNamespaces() {
@@ -76,7 +78,7 @@ public class XMLValidationSettings {
 
 	/**
 	 * Set the XML Namespaces validation settings.
-	 * 
+	 *
 	 * @param namespaces the XML Namespaces validation settings.
 	 */
 	public void setNamespaces(XMLNamespacesSettings namespaces) {
@@ -94,7 +96,7 @@ public class XMLValidationSettings {
 
 	/**
 	 * Set the XML Schema validation settings.
-	 * 
+	 *
 	 * @param schema the XML Schema validation settings.
 	 */
 	public void setSchema(XMLSchemaSettings schema) {

--- a/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/FaultTolerantTypeAdapterFactory.java
+++ b/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/FaultTolerantTypeAdapterFactory.java
@@ -1,0 +1,69 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.settings;
+
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import com.google.gson.Gson;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+
+/**
+ * Creates TypeAdapters that are wrapped in a try/catch statement, so if
+ * anything goes wrong when deserializing an object, it is set to null and not
+ * parsed.
+ *
+ */
+public class FaultTolerantTypeAdapterFactory implements TypeAdapterFactory {
+
+	private static final Logger LOGGER = Logger.getLogger(FaultTolerantTypeAdapterFactory.class.getName());
+
+	@Override
+	public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+		return wrapTypeAdapter(gson.getDelegateAdapter(this, type));
+	}
+
+	/**
+	 * Wrap the reader of the given TypeAdapter with a try/catch, and skip reading
+	 * the value if an error occurs
+	 *
+	 * @param <T>         The type that the TypeAdapter
+	 * @param typeAdapter the TypeAdapter to wrap
+	 * @return the wrapped instance of the TypeAdapter
+	 */
+	private <T> TypeAdapter<T> wrapTypeAdapter(TypeAdapter<T> typeAdapter) {
+		return new TypeAdapter<T>() {
+
+			@Override
+			public void write(JsonWriter out, T value) throws IOException {
+				typeAdapter.write(out, value);
+			}
+
+			@Override
+			public T read(JsonReader in) throws IOException {
+				try {
+					return typeAdapter.read(in);
+				} catch (Exception e) {
+					LOGGER.warning("Encountered an invalid setting. Using the default value. " + //
+							"Please check your settings for outdated or invalid settings.\n" +
+							e);
+					in.skipValue();
+					return null;
+				}
+			}
+
+		};
+	}
+
+}

--- a/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/JSONUtility.java
+++ b/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/JSONUtility.java
@@ -13,11 +13,12 @@
  *******************************************************************************/
 package org.eclipse.lemminx.utils;
 
-import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapter;
-
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.google.gson.JsonElement;
+
+import org.eclipse.lemminx.settings.FaultTolerantTypeAdapterFactory;
+import org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapter;
 
 /**
  * JSONUtility
@@ -47,6 +48,8 @@ public class JSONUtility {
 		return new GsonBuilder() //
 				// required to deserialize XMLFormattingOptions which extends FormattingOptions
 				// which uses Either
-				.registerTypeAdapterFactory(new EitherTypeAdapter.Factory());
+				.registerTypeAdapterFactory(new EitherTypeAdapter.Factory()) //
+				.registerTypeAdapterFactory(new FaultTolerantTypeAdapterFactory());
 	}
+
 }

--- a/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/settings/SettingsTest.java
+++ b/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/settings/SettingsTest.java
@@ -32,6 +32,7 @@ import org.eclipse.lemminx.extensions.contentmodel.settings.ContentModelSettings
 import org.eclipse.lemminx.extensions.contentmodel.settings.SchemaEnabled;
 import org.eclipse.lemminx.settings.capabilities.InitializationOptionsExtendedClientCapabilities;
 import org.eclipse.lemminx.utils.FilesUtils;
+import org.eclipse.lemminx.utils.JSONUtility;
 import org.eclipse.lsp4j.FormattingOptions;
 import org.eclipse.lsp4j.InitializeParams;
 import org.junit.jupiter.api.AfterEach;
@@ -248,5 +249,11 @@ public class SettingsTest {
 		assertEquals(CodeLensKind.References, clientCapabilities.getCodeLens().getCodeLensKind().getValueSet().get(0));
 		assertTrue(clientCapabilities.isActionableNotificationSupport());
 		assertTrue(clientCapabilities.isOpenSettingsCommandSupport());
+	}
+
+	@Test
+	public void oldBooleanDoesntCrashSettings() {
+		AllXMLSettings allSettings = new Gson().fromJson("{'xml': { \"validation\": { \"schema\": false}}}", AllXMLSettings.class);
+		JSONUtility.toModel(allSettings.getXml(), ContentModelSettings.class);
 	}
 }

